File: data/src/main/java/de/seuhd/campuscoffee/data/implementations/ReviewDataServiceImpl.java
package de.seuhd.campuscoffee.data.implementations;

import de.seuhd.campuscoffee.data.mapper.PosEntityMapper;
import de.seuhd.campuscoffee.data.mapper.ReviewEntityMapper;
import de.seuhd.campuscoffee.data.mapper.UserEntityMapper;
import de.seuhd.campuscoffee.data.persistence.entities.ReviewEntity;
import de.seuhd.campuscoffee.data.persistence.repositories.ReviewRepository;
import de.seuhd.campuscoffee.domain.model.objects.Pos;
import de.seuhd.campuscoffee.domain.model.objects.Review;
import de.seuhd.campuscoffee.domain.model.objects.User;
import de.seuhd.campuscoffee.domain.ports.data.ReviewDataService;
import org.jspecify.annotations.NonNull;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Set;

@Service
public class ReviewDataServiceImpl
        extends CrudDataServiceImpl<Review, ReviewEntity, ReviewRepository, Long>
        implements ReviewDataService {

    private final PosEntityMapper posEntityMapper;
    private final UserEntityMapper userEntityMapper;

    // reviews have no unique constraints, so no constraint mappings are declared
    ReviewDataServiceImpl(ReviewRepository repository, ReviewEntityMapper entityMapper,
                          PosEntityMapper posEntityMapper, UserEntityMapper userEntityMapper) {
        super(repository, entityMapper, Review.class, Set.of());
        this.posEntityMapper = posEntityMapper;
        this.userEntityMapper = userEntityMapper;
    }

    @Override
    public @NonNull List<Review> filter(@NonNull Pos pos, @NonNull Boolean approved) {
        return repository.findAllByPosAndApproved(posEntityMapper.toEntity(pos), approved)
                .stream()
                .map(mapper::fromEntity)
                .toList();
    }

    @Override
    public @NonNull List<Review> filter(@NonNull Pos pos, @NonNull User author) {
        return repository.findAllByPosAndAuthor(posEntityMapper.toEntity(pos), userEntityMapper.toEntity(author))
                .stream()
                .map(mapper::fromEntity)
                .toList();
    }
}

File: data/src/main/kotlin/de/seuhd/campuscoffee/data/implementations/ReviewDataServiceImpl.kt
package de.seuhd.campuscoffee.data.implementations

import de.seuhd.campuscoffee.data.constraints.ConstraintMapping
import de.seuhd.campuscoffee.data.mapper.PosEntityMapper
import de.seuhd.campuscoffee.data.mapper.ReviewEntityMapper
import de.seuhd.campuscoffee.data.mapper.UserEntityMapper
import de.seuhd.campuscoffee.data.persistence.entities.ReviewEntity
import de.seuhd.campuscoffee.data.persistence.repositories.ReviewRepository
import de.seuhd.campuscoffee.domain.model.objects.Pos
import de.seuhd.campuscoffee.domain.model.objects.Review
import de.seuhd.campuscoffee.domain.model.objects.User
import de.seuhd.campuscoffee.domain.ports.data.ReviewDataService
import org.springframework.stereotype.Service

/**
 * Data-layer adapter implementing the review data service port. The (pos, author) pair is unique: the
 * database constraint is the authoritative guard for the "one review per author per POS" rule, closing
 * the race the domain-level check-then-act cannot.
 */
@Service
class ReviewDataServiceImpl(
    repository: ReviewRepository,
    entityMapper: ReviewEntityMapper,
    private val posEntityMapper: PosEntityMapper,
    private val userEntityMapper: UserEntityMapper
) : CrudDataServiceImpl<Review, ReviewEntity, ReviewRepository, Long>(
        repository,
        entityMapper,
        Review::class.java,
        setOf(
            ConstraintMapping(
                { "POS ${it.pos.id}, author ${it.author.id}" },
                "pos_id/author_id",
                ReviewEntity.POS_AUTHOR_UNIQUE_CONSTRAINT
            )
        )
    ),
    ReviewDataService {
    override fun filter(
        pos: Pos,
        approved: Boolean
    ): List<Review> =
        repository
            .findAllByPosAndApproved(posEntityMapper.toEntity(pos), approved)
            .map { mapper.fromEntity(it) }

    override fun filter(
        pos: Pos,
        author: User
    ): List<Review> =
        repository
            .findAllByPosAndAuthor(posEntityMapper.toEntity(pos), userEntityMapper.toEntity(author))
            .map { mapper.fromEntity(it) }
}
